core/drivers/atmel_shdwc.c

```diff
@@ -0,0 +1,183 @@
+
+static vaddr_t shdwc_base;
+static vaddr_t mpddrc_base;
+
+bool atmel_shdwc_available(void)
+{
+	return shdwc_base != 0;
+}
+
+void __noreturn atmel_shdwc_shutdown(void)
+{
+	vaddr_t pmc_base = at91_pmc_get_base();
+
+	/* Mask exception before entering assembly which does not expect to be
+	 * interrupted.
+	 */
+	thread_mask_exceptions(THREAD_EXCP_ALL);
+
+	__atmel_shdwc_shutdown(mpddrc_base, shdwc_base, pmc_base);
+
+	/* We are going to shutdown the CPU so we will never hit this loop */
+	while (true)
+		;
+}
+
+static const unsigned long long sdwc_dbc_period[] = {
+	0, 3, 32, 512, 4096, 32768,
+};
+
+static uint32_t at91_shdwc_debouncer_value(uint32_t in_period_us)
+{
+	int i = 0;
+	int max_idx = ARRAY_SIZE(sdwc_dbc_period) - 1;
+	uint64_t period_us = 0;
+	uint64_t max_period_us = DBC_PERIOD_US(sdwc_dbc_period[max_idx]);
+
+	if (in_period_us > max_period_us) {
+		DMSG("debouncer period %u too big, reduced to %llu us",
+		     in_period_us, max_period_us);
+		return max_idx;
+	}
+
+	for (i = max_idx - 1; i > 0; i--) {
+		period_us = DBC_PERIOD_US(sdwc_dbc_period[i]);
+		if (in_period_us > period_us)
+			break;
+	}
+
+	return i + 1;
+}
+
+static uint32_t at91_shdwc_get_wakeup_input(const void *fdt, int np)
+{
+	const uint32_t *prop = NULL;
+	uint32_t wk_input_mask = 0;
+	uint32_t wuir = 0;
+	uint32_t wk_input = 0;
+	int child = 0;
+	int len = 0;
+
+	fdt_for_each_subnode(child, fdt, np) {
+		prop = fdt_getprop(fdt, child, "reg", &len);
+		if (!prop || len != sizeof(uint32_t)) {
+			DMSG("reg property is missing for node %s",
+			     fdt_get_name(fdt, child, NULL));
+			continue;
+		}
+		wk_input = fdt32_to_cpu(*prop);
+		wk_input_mask = BIT32(wk_input);
+		if (!(wk_input_mask & AT91_SHDW_WKUPEN_MASK)) {
+			DMSG("wake-up input %d out of bounds ignore", wk_input);
+			continue;
+		}
+		wuir |= wk_input_mask;
+
+		if (fdt_getprop(fdt, child, "atmel,wakeup-active-high", NULL))
+			wuir |= AT91_SHDW_WKUPT(wk_input);
+	}
+
+	return wuir;
+}
+
+static void at91_shdwc_dt_configure(const void *fdt, int np)
+{
+	const uint32_t *prop = NULL;
+	uint32_t mode = 0;
+	uint32_t tmp = 0;
+	uint32_t input = 0;
+	int len = 0;
+
+	prop = fdt_getprop(fdt, np, "debounce-delay-us", &len);
+	if (prop && len == sizeof(uint32_t)) {
+		tmp = fdt32_to_cpu(*prop);
+		mode |= AT91_SHDW_WKUPDBC(at91_shdwc_debouncer_value(tmp));
+	}
+
+	if (fdt_getprop(fdt, np, "atmel,wakeup-rtc-timer", &len))
+		mode |= AT91_SHDW_RTCWKEN;
+
+	io_write32(shdwc_base + AT91_SHDW_MR, mode);
+
+	input = at91_shdwc_get_wakeup_input(fdt, np);
+	io_write32(shdwc_base + AT91_SHDW_WUIR, input);
+}
+
+static TEE_Result atmel_shdwc_probe(const void *fdt, int node,
+				    const void *compat_data __unused)
+{
+	int ddr_node = 0;
+	size_t size = 0;
+	uint32_t ddr = AT91_DDRSDRC_MD_LPDDR2;
+
+	/* Code size of assembly code for shutdown must fit in a single
+	 * Cortex-A5 cache line which is 45 words long.
+	 */
+	assert(__atmel_shdwc_shutdown_cache_sz <= 32);
```

Comment:
Why not use a compile-time assertion in assembly like https://github.com/OP-TEE/optee_os/blob/3.15.0/core/arch/arm/kernel/entry_a64.S#L468-L478?

Reply:
Because I did not know how to do it :) I will use that.